### PR TITLE
feat(breadcrumb): add breadcrumb component

### PR DESCRIPTION
**Pull request by its-monotype**
https://user-images.githubusercontent.com/79363260/226710676-90e8963f-1dd1-4987-a358-a07bc58d1d7b.mp4

### CODE REVIEW

**Comment by tonyxiao** [comment]
@its-monotype Can we use `asChild` rather than `as`? This way we can do things like passing in a next.js `Link` component with typed hrefs for example.

**Comment by its-monotype** [comment]
> @its-monotype Can we use `asChild` rather than `as`? This way we can do things like passing in a next.js `Link` component with typed hrefs for example. 

Yeah, great idea, would look into it

**Comment by its-monotype** [comment]
@tonyxiao Inside BreadcrumbLink, we currently replace the current element with a <span> if it represents the current page. While the "asChild" approach you mentioned has its merits, it poses a challenge when it comes to replacing or modifying the slotted child, as well as removing the href attribute from it. I don't have any ideas on how to solve this so far, and would greatly appreciate any help in finding a solution.

**Review by vermeerenmaxime — APPROVED** [review]

**Review by Beor18 — APPROVED** [review]

**Comment by focux** [comment]
I found this PR when looking for a Breadcrumb component for shadcn/ui. Awesome work @its-monotype! This would be a great addition to the library, I hope this PR doesn't go stale.

**Comment by Zac-Zajdel** [comment]
This repo has made me SOOOOO productive on my side projects and I was reaching for a breadcrumb component but saw it wasn't done yet.

This is working fantastic in my application right now. Thank you @its-monotype for implementing this.

+1 to getting this in as an oficially supported component!

Appreciate what you have done for the community @shadcn

**Comment by mmestiyak** [comment]
anyone working on it!

**Comment by karam-khanna** [comment]
Looks really good visually! Would +1 a previous request to support using Next Links if this PR gets picked back up again. Makes a pretty big speed difference and is a bit jarring to the user if the rest of the app is using Next Links but these aren't.

**Comment by karam-khanna** [comment]
In response to my prev comment, if anyone else is looking for a next link breadcrumb, can use the breadcrumb in this PR with the following updated BreadcrumbLink:

`export const BreadcrumbLink = React.forwardRef<
        HTMLAnchorElement,
        BreadcrumbLinkProps
>(
        ({className, as: asComp, isCurrentPage, ...props}, forwardedRef) => {
            if (isCurrentPage) {
                return (
                        <span
                                className={cn(
                                        "text-sm font-medium underline-offset-4 aria-[current]:opacity-60",
                                        className
                                )}
                                aria-current="page"
                                {...props}
                                ref={forwardedRef}
                        />
                );
            }

            return (
                    <Link href={props.href as string} passHref {...props}
                          className={cn(
                                  "text-sm font-medium underline-offset-4 [&:not([aria-current])]:hover:underline",
                                  className
                          )}
                          ref={forwardedRef}>
                    </Link>
            );
        }
)`

great work @its-monotype regardless

**Comment by Talent30** [comment]
@shadcn Could review this pr if you have time please, much appreciated :)

**Comment by shadcn** [comment]
@its-monotype I know this is a big ask (since this was blocked by me) but any chance you can reroll this against `main`? If not, I'll take a stab at it during the week. Thank you.

**Comment by matbrgz** [comment]
@its-monotype this feature could be awesome, is there possible to do what @shadcn said.

File: apps/www/components/ui/breadcrumb.tsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+   */
+  addSeparator?: boolean
+}
+
+export const Breadcrumb = React.forwardRef<HTMLElement, BreadcrumbProps>(
+  (
+    {
+      children,
+      className,
+      separator = <ChevronRight className="h-4 w-4" />,
```
Comment (A7med3bdulBaset) [inline apps/www/components/ui/breadcrumb.tsx:21]:
```suggestion
      separator = <ChevronRight className="h-4 w-4 rtl:rotate-180" />,
```

File: apps/www/components/ui/breadcrumb.tsx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+      </li>
+    )
+  }
+)
+BreadcrumbItem.displayName = "BreadcrumbItem"
+
+export interface BreadcrumbLinkProps
+  extends React.ComponentPropsWithoutRef<"a">,
+    Pick<BreadcrumbItemProps, "isCurrentPage"> {
+  as?: React.ElementType
```
Comment (A7med3bdulBaset) [inline apps/www/components/ui/breadcrumb.tsx:109]: Can you use `@radix-ui/react-slot`'s `asChild to keep the same design system?

File: apps/www/components/ui/breadcrumb.tsx [hunk 3]
```diff
@@ -0,0 +1,147 @@
+import * as React from "react"
+import { ChevronRight } from "lucide-react"
+
+import { cn, getValidChildren } from "@/lib/utils"
+
+export interface BreadcrumbProps extends React.ComponentPropsWithoutRef<"nav"> {
```
Comment (A7med3bdulBaset) [inline apps/www/components/ui/breadcrumb.tsx:6]: Is there a reason for exporting the types?

**Comment by Balance8** [comment]
bumping this!

**Comment by shadcn** [comment]
@Balance8 I'll make some updates to the component API and merge. Thank you.

**Comment by shakedlokits** [comment]
We're excited and waiting for it! thank you!

**Comment by SirawichDev** [comment]
any updates quite good to have this component in Shadcn :3

**Comment by birobirobiro** [comment]
I need this component!

**Comment by JonasDoesThings** [comment]
Ofc I am also looking forward to an official integration, but to everyone waiting impatiently: just copy the component manually into your project for now 🤷

**Comment by birobirobiro** [comment]
> Ofc I am also looking forward to an official integration, but to everyone waiting impatiently: just copy the component manually into your project for now 🤷

That's exactly what I did, I'll update later when they merge

File: apps/www/components/examples/breadcrumb/with-routing-library.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  BreadcrumbItem,
+  BreadcrumbLink,
+} from "@/components/ui/breadcrumb"
+import { Icons } from "@/components/icons"
+
+export function BreadcrumbWithRoutingLibrary() {
+  return (
+    <Breadcrumb>
+      <BreadcrumbItem>
+        <BreadcrumbLink as={Link} className="flex items-center gap-2" href="/">
```
Comment (Dude29) [inline apps/www/components/examples/breadcrumb/with-routing-library.tsx:15]: Would this also work with [react-router Links](https://reactrouter.com/en/main/components/link)?
If yes would it be possible for you to add another example file demonstrating this?

**Comment by mohab-sameh** [comment]
Awesome work @its-monotype. Will be super helpful when it gets merged.

**Comment by NicolasStuff** [comment]
Awesome! I also need this component, hope it will be merge soon ⚡️

**Review by Beor18 — APPROVED** [review]

**Comment by rickylabs** [comment]
Any status why it's still not meged ?

**Comment by premiare** [comment]
> Any status why it's still not meged ?

It's currently in the milestones https://github.com/shadcn-ui/ui/milestone/1

File: apps/www/components/ui/breadcrumb.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      className={cn(
+        "text-sm font-medium underline-offset-4 aria-[current]:opacity-60 [&:not([aria-current])]:hover:underline",
+        className
+      )}
+      aria-current={isCurrentPage ? "page" : undefined}
+      {...props}
+      ref={forwardedRef}
+    />
+  )
+})
```
Comment (m0a) [inline apps/www/components/ui/breadcrumb.tsx:129]:
```suggestion
export type BreadcrumbLinkProps<T extends React.ElementType = "a"> = Pick<BreadcrumbItemProps, "isCurrentPage"> & {
    as?: T;
} & React.ComponentPropsWithoutRef<T>;

export const BreadcrumbLink = React.forwardRef(
    <T extends React.ElementType = "a">(
      { isCurrentPage, as, className, ...props }: BreadcrumbLinkProps<T> & { as?: T },
      ref: React.ForwardedRef<HTMLAnchorElement | HTMLSpanElement>
    ) => {
      const ElementType: React.ElementType = isCurrentPage ? "span" : as || "a";
  
      return (
        <ElementType
          {...props}
          ref={ref}
          className={cn(
            "text-sm font-medium underline-offset-4 aria-[current]:opacity-60 [&:not([aria-current])]:hover:underline",
            className
          )}
          aria-current={isCurrentPage ? "page" : undefined}
        />
      );
    }
  );
```

This change modifies the BreadcrumbLink component from being specialized for ``<a>`` tag props to accepting any tag. This is particularly useful, for example, when using the React Router's Link component, which requires a to parameter to specify the destination. The previous implementation did not support this, but with this update, it now accommodates such cases.

**Review by m0a — CHANGES_REQUESTED** [review]

**Comment by shadcn** [comment]
Hi everyone, I've made some updates to the `breadcrumb` component. Let me know if you have any feedback.

1. Reworked the component a bit so that they are more composable.
2. Added a `<BreadcrumbEllipsis />` component to display collapsed state.
3. Refactored to use `asChild` prop for custom router `<Link />`.
4. Added examples for _custom separator, dropdown and responsive with drawer_.

👉 Demo: https://ui-git-fork-its-monotype-feat-breadcrumb-shadcn-pro.vercel.app/docs/components/breadcrumb

hSaOjqpN

File: apps/www/content/docs/components/breadcrumb.mdx [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+---
+
+### Collapsed
+
+We provide a `<BreadcrumbEllipsis />` component to show a collapsed state when the breadcrumb is too long.
+
+<ComponentPreview name="breadcrumb-ellipsis" className="[&_.preview]:p-2" />
+
+```tsx showLineNumbers {1,13-15}
```
Comment (flixlix) [inline apps/www/content/docs/components/breadcrumb.mdx:141]: shouldn't this be highlighting line 1 & 9?
Like this:
` ```tsx showLineNumbers {1,9}`

Comment (shadcn) [reply apps/www/content/docs/components/breadcrumb.mdx]: @flixlix Fixed. Thanks.

**Comment by lr0pb** [comment]
Drawer example have no code for it, and probably all other examples work not as expected - code is below the preview, not in the `Code` section

**Comment by premiare** [comment]
> Drawer example have no code for it, and probably all other examples work not as expected - code is below the preview, not in the `Code` section

Custom separator as an example:

`Preview` shows the code in context (related to that example)

`Code` shows all the code for that example

Drawer and Dialog have very simple implementations:

`Drawer`
```tsx
export const BreadcrumbDrawer = () => {
  return (
    <Drawer>
      <DrawerTrigger>Drawer</DrawerTrigger>
      <DrawerContent className="h-[400px]">
        <DrawerHeader>
          <Breadcrumb>
            <BreadcrumbList>
              <BreadcrumbItem>
                <BreadcrumbLink href="/">Home</BreadcrumbLink>
              </BreadcrumbItem>
              <BreadcrumbSeparator />
              <BreadcrumbItem>
                <BreadcrumbLink href="/">Dashboard</BreadcrumbLink>
              </BreadcrumbItem>
              <BreadcrumbSeparator />

              <BreadcrumbItem>
                <BreadcrumbLink href="/">Users</BreadcrumbLink>
              </BreadcrumbItem>
              <BreadcrumbSeparator />

              <BreadcrumbItem>
                <BreadcrumbLink href="/">John Doe</BreadcrumbLink>
              </BreadcrumbItem>
            </BreadcrumbList>
          </Breadcrumb>
        </DrawerHeader>
      </DrawerContent>
    </Drawer>
  );
};
```
image

`Dialog`

```tsx
export const BreadcrumbDialog = () => {
  return (
    <Dialog>
      <DialogTrigger>Breadcrumb</DialogTrigger>
      <DialogContent>
        <DialogHeader>
          <Breadcrumb>
            <BreadcrumbList>
              <BreadcrumbItem>
                <BreadcrumbLink href="/">Home</BreadcrumbLink>
              </BreadcrumbItem>
              <BreadcrumbSeparator />
              <BreadcrumbItem>
                <BreadcrumbLink href="/">Dashboard</BreadcrumbLink>
              </BreadcrumbItem>
              <BreadcrumbSeparator />

              <BreadcrumbItem>
                <BreadcrumbLink href="/">Users</BreadcrumbLink>
              </BreadcrumbItem>
              <BreadcrumbSeparator />

              <BreadcrumbItem>
                <BreadcrumbLink href="/">John Doe</BreadcrumbLink>
              </BreadcrumbItem>
            </BreadcrumbList>
          </Breadcrumb>
        </DialogHeader>
        Dialog content
      </DialogContent>
    </Dialog>
  );
};
```
image

File: apps/www/registry/default/ui/breadcrumb.tsx [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    role="presentation"
+    aria-hidden="true"
+    className={cn("flex h-9 w-9 items-center justify-center", className)}
+    {...props}
+  >
+    <MoreHorizontal className="h-4 w-4" />
+    <span className="sr-only">More</span>
+  </span>
+)
+BreadcrumbEllipsis.displayName = "BreadcrumbElipssis"
```
Comment (pitops) [inline apps/www/registry/default/ui/breadcrumb.tsx:105]: @shadcn there is a typo here

**Pull request merged**